massa-versioning-worker/src/versioning.rs

```diff
@@ -448,31 +446,70 @@ impl MipStore {
         // Defined == Not yet ready to announce
         // Active == current version
         store
-            .0
+            .store
             .iter()
             .rev()
             .find_map(|(k, v)| {
                 matches!(
-                    &v.inner,
+                    &v.state,
                     &ComponentState::Started(_) | &ComponentState::LockedIn(_)
                 )
                 .then_some(k.version)
             })
             .unwrap_or(0)
     }
+
+    pub fn update_network_version_stats(
+        &mut self,
+        slot_timestamp: MassaTime,
+        network_versions: Option<(u32, u32)>,
+    ) {
+        let mut lock = self.0.write();
+        lock.update_network_version_stats(slot_timestamp, network_versions);
+    }
 }
 
-impl<const N: usize> TryFrom<[(MipInfo, MipState); N]> for MipStore {
+impl<const N: usize> TryFrom<([(MipInfo, MipState); N], MipStatsConfig)> for MipStore {
     type Error = ();
 
-    fn try_from(value: [(MipInfo, MipState); N]) -> Result<Self, Self::Error> {
-        MipStoreRaw::try_from(value).map(|store_raw| Self(Arc::new(RwLock::new(store_raw))))
+    fn try_from(
+        (value, cfg): ([(MipInfo, MipState); N], MipStatsConfig),
+    ) -> Result<Self, Self::Error> {
+        MipStoreRaw::try_from((value, cfg)).map(|store_raw| Self(Arc::new(RwLock::new(store_raw))))
+    }
+}
+
+/// Statistics in MipStoreRaw
+#[derive(Debug, Clone, PartialEq)]
+pub struct MipStatsConfig {
+    pub block_count_considered: usize,
+    pub counters_max: usize,
+}
+
+/// In order for a MIP to be accepted, we compute stats about other node 'network' version announcement
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct MipStoreStats {
+    pub(crate) config: MipStatsConfig,
+    pub(crate) latest_announcements: VecDeque<u32>,
+    pub(crate) network_version_counters: BTreeMap<u32, u64>,
+}
+
+impl MipStoreStats {
+    pub(crate) fn new(config: MipStatsConfig) -> Self {
+        Self {
+            config: config.clone(),
+            latest_announcements: VecDeque::with_capacity(config.block_count_considered),
+            network_version_counters: Default::default(),
+        }
     }
 }
 
 /// Store of all versioning info
-#[derive(Debug, Clone, Default, PartialEq)]
-pub struct MipStoreRaw(pub(crate) BTreeMap<MipInfo, MipState>);
+#[derive(Debug, Clone, PartialEq)]
+pub struct MipStoreRaw {
+    pub(crate) store: BTreeMap<MipInfo, MipState>,
+    pub(crate) stats: MipStoreStats,
+}
 
 impl MipStoreRaw {
     /// Update our store with another (usually after a bootstrap where we received another store)
```

Comment:
Maybe add a TODO flag to also update the MipStoreStats?
And if it's too big, split it in two methods, to update the store and the stats separately.

Reply:
From what I see, the update_with will replace the stats. Not sure if there are checks that can be done?

Reply:
Mhh probably yes. For checks I think it depends if the config is included in the stats or not (see my comment below), if it's not we can check that the buffer length and max counter values are consistant with what we receive.